src/groups/orthogonal.jl

```diff
@@ -0,0 +1,154 @@
+    sinθ, cosθ = sincos(θ)
+    @inbounds begin
+        q[1, 1] = cosθ
+        q[2, 1] = sinθ
+        q[1, 2] = -sinθ
+        q[2, 2] = cosθ
+    end
+    return q
+end
+function exp_lie!(::Orthogonal{3}, q, X)
+    θ = norm(X) / sqrt(2)
+    if θ ≈ 0
+        a = 1 - θ^2 / 6
+        b = θ / 2
+    else
+        a = sin(θ) / θ
+        b = (1 - cos(θ)) / θ^2
+    end
+    copyto!(q, I)
+    q .+= a .* X
+    mul!(q, X, X, b, true)
+    return q
+end
+function exp_lie!(::Orthogonal{4}, q, X)
+    T = eltype(X)
+    α, β = angles_4d_skew_sym_matrix(X)
+    sinα, cosα = sincos(α)
+    sinβ, cosβ = sincos(β)
+    α² = α^2
+    β² = β^2
+    Δ = β² - α²
+    if !isapprox(Δ, 0; atol=1e-6)  # Case α > β ≥ 0
+        sincα = sinα / α
+        sincβ = β == 0 ? one(T) : sinβ / β
+        a₀ = (β² * cosα - α² * cosβ) / Δ
+        a₁ = (β² * sincα - α² * sincβ) / Δ
+        a₂ = (cosα - cosβ) / Δ
+        a₃ = (sincα - sincβ) / Δ
+    elseif α == 0 # Case α = β = 0
+        a₀ = a₁ = one(T)
+        a₂ = inv(T(2))
+        a₃ = inv(T(6))
+    else  # Case α ⪆ β ≥ 0, α ≠ 0
+        sincα = sinα / α
+        r = β / α
+        c = 1 / (1 + r)
+        d = α * (α - β) / 2
+        if α < 1e-2
+            e = evalpoly(α², (inv(T(3)), inv(T(-30)), inv(T(840)), inv(T(-45360))))
+        else
+            e = (sincα - cosα) / α²
+        end
+        a₀ = (α * sinα + (1 + r - d) * cosα) * c
+        a₁ = ((3 - d) * sincα - (2 - r) * cosα) * c
+        a₂ = (sincα - (1 - r) / 2 * cosα) * c
+        a₃ = (e + (1 - r) * (e - sincα / 2)) * c
+    end
+
+    X² = X * X
+    copyto!(q, a₀ * I)
+    q .+= a₂ .* X²
+    mul!(q, X², X, a₃, true)
+    return q
+end
+
+function log_lie!(G::Orthogonal, X::AbstractMatrix, q::AbstractMatrix)
+    log_safe!(X, q)
+    return project!(G, X, Identity(G, q), X)
+end
+function log_lie!(G::Orthogonal{2}, X::AbstractMatrix, q::AbstractMatrix)
+    @assert size(q) == (2, 2)
+    @inbounds θ = atan(q[2, 1], q[1, 1])
+    return get_vector!(G, X, Identity(G, q), θ, DefaultOrthogonalBasis())
+end
+function log_lie!(G::Orthogonal{3}, X::AbstractMatrix, q::AbstractMatrix)
+    e = Identity(G, q)
+    cosθ = (tr(q) - 1) / 2
+    if cosθ ≈ -1
+        eig = eigen_safe(q)
+        ival = findfirst(λ -> isapprox(λ, 1), eig.values)
+        inds = SVector{3}(1:3)
+        ax = eig.vectors[inds, ival]
+        return get_vector!(G, X, e, π * ax, DefaultOrthogonalBasis())
+    end
+    X .= q ./ usinc_from_cos(cosθ)
+    return project!(G, X, e, X)
+end
+function log_lie!(G::Orthogonal{4}, X::AbstractMatrix, q::AbstractMatrix)
+    cosα, cosβ = cos_angles_4d_rotation_matrix(q)
+    α = acos(clamp(cosα, -1, 1))
+    β = acos(clamp(cosβ, -1, 1))
+    if α ≈ π && β ≈ 0
+        A² = Symmetric((q - I) ./ 2)
+        P = eigvecs(A²)
+        E = similar(q)
+        fill!(E, 0)
+        α = acos(clamp(cosα, -1, 1))
+        E[2, 1] = -α
+        E[1, 2] = α
+        copyto!(X, P * E * transpose(P))
+    else
+        log_safe!(X, q)
```

Comment:
I am not sure what is done here, but it currently fails a test, since points like `Diagonal([-1.0, -1.0, -1.0, 1.0})`are valid here but hit this line _and_ have a determinant of -1. So where is this formula from? Is it correct? What do now? @sethaxen ?

Reply:
Well, as far as I know there can be no `log_lie` for point with determinant -1. `Diagonal([-1.0, -1.0, -1.0, 1.0})` lies in a different connected component. I guess we should throw an error?

Reply:
I just discussed this with Seth and you might be right (then one could just add a `is_point(SpecialOrthogonal,...)` there?) Let's see, I will wait for one feedback, since currently on SO(n) and O(n) (which then have the same log/exp lie) we currently have different implementations.

Reply:
right, we would just have to check the determinant, that's right.

Reply:
Computing the determinant is ~1% the cost of computing the matrix logarithm for an orthogonal matrix, so this seems a good way to go.

Reply:
Perfect, will add that. This function (Orthogonal(4) log lie) seems to be the only test failing for now, doing exp-then-log yields

```
isapprox(Orthogonal(4), Identity(MultiplicationOperation), [0.0 -0.2853981633974483 0.0 0.0; 0.2853981633974483 0.0 0.0 0.0; 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0], [0.0 1.0 0.0 0.0; -1.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0])
```

(where the second one is the one I start with and after exp-then-log I have the first one), I will for now add that as a broken test to check for further test coverage.

Reply:
Ah, I see now there's an error in the exponential map for `Orthogonal{4}`. I never include the `a₁` term. Just revert to the `Rotations{4}` version for now.

Reply:
Thanks I will check – the `Orthogonal{4}` function is really just the one from the old branch/PR.

Reply:
It works fine, I just have no clue how to hit the special if cases in code coverage.

Reply:
see https://app.codecov.io/gh/JuliaManifolds/Manifolds.jl/compare/499/tree/src/groups/general_unitary_groups.jl

Reply:
> Thanks I will check – the `Orthogonal{4}` function is really just the one from the old branch/PR.

Which in turn is a slight tweak of `Rotations{4}` to remove an allocation or 2.

> It works fine

No, it is incorrect as noted above. Please revert to the `Rotations{4}` implementation for now.

> I just have no clue how to hit the special if cases in code coverage

See the `Rotations{4}` tests: 
https://github.com/JuliaManifolds/Manifolds.jl/blob/9a76bad003ef14339c1f61cc33e607efd990e198/test/manifolds/rotations.jl#L161-L184

Reply:
Hm? I just copied for the last commit the code from Rot4.

Sure I will adapt these tests then (but they should be run anyways on that code base now.

Reply:
so in short - I am unsure why this exp_lie / group_exp has a q https://github.com/JuliaManifolds/Manifolds.jl/blob/38e20ead10af77ae241002fb3abaf309d9c17ea0/src/groups/orthogonal.jl#L169

Reply:
ah no, I am just stupid.

Reply:
What was the issue in the end? :sweat_smile:

Reply:
> Ah, I see now there's an error in the exponential map for `Orthogonal{4}`. I never include the `a₁` term. Just revert to the `Rotations{4}` version for now.

so rot4 for example does not have a group log?

Reply:
I pushed the current state which fails O(n) still. I am a little lost which is now the right implementation for exp_lie (formerly group_exp) and its log for n=4

Reply:
Use this for `exp`:
https://github.com/JuliaManifolds/Manifolds.jl/blob/9a76bad003ef14339c1f61cc33e607efd990e198/src/manifolds/Rotations.jl#L203-L243
and this for `log`:
https://github.com/JuliaManifolds/Manifolds.jl/blob/9a76bad003ef14339c1f61cc33e607efd990e198/src/manifolds/Rotations.jl#L604-L624

Reply:
Just replace `p` with `I`, and you have the `exp_lie`. but you can probably remove an allocation by not multiplying by `I`.

Reply:
Ok, I will check, my problem was exactly that, together with the exp/log (end exp/log_lie) I had 6 implementations, (Rotations, SpecialOrthogonal, Orthogonal) for each (exp and log) – so I got totally lost which one to take now.

Reply:
hm, but now line 241 includes a1 again which you said should not be in there? That is also a difference between the one `exp_lie I have, i.e. 

```
    copyto!(q, a₀ * I)
    q .+= a₂ .* X²
    mul!(q, X², X, a₃, true)
```

and the other one which is (in exp, so this is multiplied by p afterwards

```
    pinvq = a₀ * I + a₁ .* X .+ a₂ .* X² .+ a₃ .* X³
```

which one is now correct?

Reply:
My question was just which one was correct ;) I am further fiddling around (lost all afternoon anyways already to this). It was really tricky with soooooooooooo many implementations basically doing the same.

Reply:
*phew* I think I finally got it and reduced the code redundancy here, also put all non-group stuff into the matrix/set definitions and the group/lie definitions into the group setting.